spec/requests/profile_spec.rb

```diff
@@ -0,0 +1,164 @@
+
+      it "returns a no_content status" do
+        expect(response).to have_http_status(:no_content)
+      end
+    end
+  end
+
+  describe "#add_ban" do
+    let(:user) { create(:user2) }
+
+    before(:each) { Devise.mailer.deliveries.clear }
+
+    context "when signed in user is not an admin" do
+      let(:non_admin_user) { create(:user1) }
+      before { sign_in non_admin_user }
+
+      it "cannot ban user" do
+        post add_ban_profile_index_path, params: {user_id: user.id}
+        expect(response).to have_http_status(204)
+        expect(Devise.mailer.deliveries.count).to eq(0)
+      end
+    end
+
+    context "when signed in user is an admin" do
+      let(:admin_user) { create(:user1, admin: true) }
+      before { sign_in admin_user }
+
+      context "when user exists" do
+        it "bans the user" do
+          expect { post add_ban_profile_index_path, params: {user_id: user.id} }
+            .to change(Devise.mailer.deliveries, :count).by 1
+          expect(response).to redirect_to(admin_dashboard_path)
+          expect(flash[:notice]).to eq("#{user.name} has been banned")
+        end
+      end
+
+      context "when user does not exist" do
+        it "does not ban the user" do
+          expect { post add_ban_profile_index_path, params: {user_id: -1} }
+            .not_to change(Devise.mailer.deliveries, :count).from(0)
+          expect(response).to redirect_to(admin_dashboard_path)
+          expect(flash[:alert]).to eq("Could not ban -1")
+        end
+      end
+    end
+  end
+
+  describe "#remove_ban" do
+    let(:banned_user) { create(:user2, banned: true) }
+
+    before(:each) { Devise.mailer.deliveries.clear }
+
+    context "when signed in user is not an admin" do
+      let(:nonadmin_user) { create(:user1) }
+      before { sign_in nonadmin_user }
+
+      it "cannot ban user" do
+        expect { post remove_ban_profile_index_path, params: {user_id: banned_user.id} }
+          .not_to change(Devise.mailer.deliveries, :count).from(0)
+        expect(response).to have_http_status(204)
+      end
+    end
+
+    context "when signed in user is an admin" do
+      let(:admin_user) { create(:user1, admin: true) }
+      before { sign_in admin_user }
+
+      context "when user exists" do
+        it "removes ban" do
+          expect { post remove_ban_profile_index_path, params: {user_id: banned_user.id} }
+            .to change(Devise.mailer.deliveries, :count).by(1)
+          expect(response).to redirect_to(admin_dashboard_path)
+          expect(flash[:notice]).to eq("Ban removed on #{banned_user.name}")
+        end
+      end
+
+      context "when user does not exist" do
+        it "does not remove ban" do
+          expect { post remove_ban_profile_index_path, params: {user_id: -1} }
+            .not_to change(Devise.mailer.deliveries, :count).from(0)
+          expect(response).to redirect_to(admin_dashboard_path)
+          expect(flash[:alert]).to eq("Could not remove ban on -1")
```

Comment:
```suggestion
          expect(flash[:alert]).to include("Could not remove ban")
```

Nitpick on all the flash alert equals, but you can also use `include` and just put part of the string as well just to make the less likely to break if its a simple verbiage change on the string. Overall, it's not a big deal. I consider the tests ok.

Reply:
That works for me.